Added PhysX integration

AtlantiaKing opened the pull request:
Added physics integration with a PhysX service.

Features:
- BoxCollider, CapsuleCollider and SphereCollider components that can be triggers
- Collision & Trigger callbacks inside the collider components
- Physics Materials
- Rigidbody component with mass, forces, torque, constraints and can be set to kinematic
- Rigidbody has its own SetPosition & SetRotation functions to move a physics actor since Transform doesn't work anymore

AtlantiaKing committed:
Added IPhysics and PhysXEngine setup

Added IPhysics interface with DefaultPhysics class
Added PhysXEngine which sets up top-level PhysX objects at construction and releases them on destruction
Added physics to the servicelocator from the engine
Fixed additional physX libraries not being included by cmake
Fixed physX header files not being included by cmake

AtlantiaKing committed:
Added physX scene wrapper

Added IPhysicsScene interface and added PhysXScene class that derives from the interface
Added CreateScene function to the IPhysics and PhysXEngine classes
Scene constructor calls CreateScene function from the IPhysics interface

AtlantiaKing committed:
Added boxcollider & rigidbody

Added cube obj
Added PhysicsSync class with static function to sync from physX to transform component
Added IPhysicsObject interface and PhysXObject classe which is a wrapper around a rigidactor
Added IShape interface and PhysXShapes file which will hold all the PhysX shapes as different classes
Modified physics engine to be able to add actors and simulate the scene
Added physics to the main game loop
Added BoxCollider & Rigidbody components to the main engine

AtlantiaKing committed:
Added sphere collider

Added sphere shape to PhysXShapes
Added SphereCollider component

AtlantiaKing committed:
Added SetRadius and SetSize to shapes

Added SetSize and SetRadius functions to the IShape class
Added SetSize function to the BoxCollider component
Added SetRadius function to the SphereCollider component
Added sphere obj

AtlantiaKing committed:
Rigidbody now groups together with child colliders

Added base collider component, BoxCollider and SphereCollider now derive from this Collider component
Rigidbody now makes sure child colliders have the same rigid actor
Transforms of static colliders are now no longer being updated by the physics system, instead the physics bodies of static colliders are being updated by the transforms
Fixed compile error in the GameObject GetComponentInParent class
Fixed PhysicsObjects not being deleted from the scene properly when they are destroyed

AtlantiaKing committed:
Added COM calculations

When a shape is added or removed from a rigidbody, the center of mass (COM) is now being recalculated
Added GetVolume and GetRelativePosition methods to the IShape interface and its derived classes

AtlantiaKing committed:
Fixed mass calculations and added SetMass function

Calcualte CenterOfMass functino in the PhysXObject class now uses the PxRigidBodyExt::setMassAndUpdateIntertia function, which sets more parameters then just the COM for correct behaviour.
Moved Rigidbody from inside the IPhysicsObject interface to its own file
Added SetMass functino to Rigidbody component

AtlantiaKing committed:
Fixed COM being calculated before shapes are applied to an object

AtlantiaKing committed:
Fixed wrong quaterion/euler calculations in transform

Replaced the calculations from the glm library to convert quaterions to euler and euler to quaterions with my own implementation.

AtlantiaKing committed:
Fixed camera rotation not working when looking in certain directions

Removed the yaw rotation on the parent transform of the camera object. Each object should only do one rotation, the parent does yaw, the camera itself does pitch.

AtlantiaKing committed:
Added Quaternion class with helper functions

Moved all the quaternion/euler angles calculations to a Quaternion class inside the utils engine, it is non constructable and has some static helper functions to correctly convert from euler to quaternion and from quaternion to euler

AtlantiaKing committed:
Added overloads for functions that take a glm::vec3

Added overloads for functions like SetSize and SetVelocity to also take the seperate floats instead of a glm::vec3 object

AtlantiaKing committed:
Added VLD to game project & fixed memory leaks

Added Visual Leak Detector to the 3rdParty folder and included it only into the game project
Included vld into main.cpp of the game project
Fixed memory leaks in PhysXEngine.cpp

AtlantiaKing committed:
Fixed transform/rigidbody issues + added more rigidbody functions and capsule collider

Added rule of 5 to IPhysXShape
Added PhysXCapsuleShape
Added GetPosition and GetRotation to IPhysicsObject and PhysXObject
Fixed order of PhysXObject updates (Rigidbody was being updated after the transform was updated)
Added SetPosition, Translate, SetRotation and Rotate functions to the Rigidbody component
Fixed direction vectors of child transform not being updated after rotation has changed in parent

AtlantiaKing committed:
Added physics materials

Added IPhysicsMaterial interface and PhysXMaterial class
Added CreateMaterial function to the IPhysics interface and PhysXEngine class
Added SetMaterial function to the Collider base class
Colliders that get created without a physics collider set will be created using a default material created by the physics engine

AtlantiaKing committed:
Added collision callbacks

Added PhysXSimulationCallbacks & PhysXSimulationFilterShader which are owned by the PhysXEngine and passed to the PxScene on creation. PhysXSimulationFilterShader currently lets all the collisions succeed and notify
A pointer to the component collider is now stored inside the userdata of the corresponding PxShape
Added static OnCollision method to the PhysicsSync class to handle incoming collision events from the physics engine
Added Subject to the Collider component which gets called when that collider receives a collision event
Added PrintCollision component which subscribes to the OnCollision subject of the collider on the same gameobject and prints the names of the gameobjects that are colliding

AtlantiaKing committed:
Added AddForce and AddTorque to the rigidbody component

Moved temporary rigidbody creation to a CheckExistence function instead of copy pasting the code in each function
Added Force struct, which stores the value, if it is a torque or not, and the force mode
Added AddForce and AddTorque to the Rigidbody class and the Rigidbody component
PhysXObject now applies forces stored in the rigidbody class
Added ApplyForces component to test force functions

AtlantiaKing committed:
Added SetTrigger function to colliders

Moved some logic to IPhysXShape and added SetTrigger function to IShape and IPhysXShape
Added SetTrigger function to base collider component
Added trigger checking to the PhysXSimulationFilterShader

AtlantiaKing committed:
Added more collision & trigger events

Renamed OnCollision to OnCollisionEnter, Added OnCollisionStay and OnCollisionExit
Added PhysXSimulationfilterCallback class which gives us more control
Added setup for OnTriggerEnter, OnTriggerStay and OnTriggerExit but the SimulationfilterCallback isn't being called for trigger shapes yet, that's the next step.

AtlantiaKing committed:
Finalized OnTrigger events

SimulationFilterShader now returns default filter flag for triggers
Moved SimulationEvent and SimulationEventType from PhysXSimulationFilterCallback.h to PhysXSimulationData.h and SimulationEventData.h
Moved SimulationPair from PhysXSimulationFilterCallback.h to PhysXSimulationData.h
Trigger events are now being handled by the PhysxSimulationCallbacks class instead of by the PhysXSimulationFilterCallback class

AtlantiaKing committed:
Added constraints to rigidbody

AtlantiaKing committed:
SetSize/SetRadius on colliders can now be called during runtime

Added RescaleShape pure virtual method to the base collider and added overrides to all the collider components
RescaleShape is now being called when scale has changed on the transform of the component
Added capsule.obj
Renamed Observer and Subject to TObserver and TSubject and added a non-templated Observer & Subject class

LeeVangraefschepe commented on LeapEngine/Components/Physics/BoxCollider.h:
```diff
@@ -0,0 +1,32 @@
+
+#include <vec3.hpp>
+
+#include <Observer.h>
```
Remove unused include in header file
Base class uses it (Collider) but BoxCollider doesnt

LeeVangraefschepe commented on LeapEngine/Components/Physics/BoxCollider.cpp:
```diff
@@ -0,0 +1,44 @@
+#include "../../SceneGraph/GameObject.h"
+
+#include "../Transform/Transform.h"
+#include "Rigidbody.h"
```
Unused include

LeeVangraefschepe commented on LeapEngine/Components/Physics/Collider.cpp:
```diff
@@ -0,0 +1,118 @@
+
+void leap::Collider::SetMaterial(std::shared_ptr<physics::IPhysicsMaterial> pMaterial)
+{
+	m_pMaterial = pMaterial;
```
Passing copy from std::shared_ptr. Should be const ref

LeeVangraefschepe commented on LeapEngine/Components/Physics/Rigidbody.h:
```diff
@@ -0,0 +1,59 @@
+		Rigidbody& operator=(Rigidbody&& other) = delete;
+
+		void SetKinematic(bool isKinematic);
+		void SetVelocity(const glm::vec3& velocity);
```
Set angular velocity is missing

LeeVangraefschepe commented on LeapEngine/Utils/PhysicsSync.cpp:
```diff
@@ -0,0 +1,79 @@
+
+leap::PhysicsSync::ColliderPair leap::PhysicsSync::GetColliders(const physics::CollisionData& collision)
+{
+	Collider* pFirstCollider{ reinterpret_cast<Collider*>(collision.pFirst) };
```
Use static cast

AtlantiaKing committed:
Applied code review feedback

Removed redundant includes
shared_ptr parameter in the SetMaterial function of the Collider class is now passed by reference instead of copy
GetColliders function in the PhysicsSync class now uses static_cast to convert void* to Collider* instead of reinterpret_cast

AtlantiaKing committed:
Added SetAngularVelocity to Rigidbody component

Added SetAngularVelocity and GetAngularVelocity functions and AngularVelocity dirty flag to the physics Rigidbody data class.
Added SetAngularVelocity to the Rigidbody component

AtlantiaKing committed:
Added get functions for Rigidbody component

Added GetVelocity, GetAngularVelocity, GetMass, IsConstraint and IsKinematic functions to the rigidbody component

AtlantiaKing committed:
Added debug rendering to physx

Added SetEnabledDebugDrawing and GetDebugDrawings methods to IPhysics, IPhysicsScene, PhysXEngine and PhysXScene classes
Fixed bug in release mode where the physics engine would use dangling references from the get function from the physicssync class. This function now returns copies instead of references.
Fixed deltatime being bigger then was requested. Removed this_thread::sleep_for call and replaced it with a blocking while loop
Added DrawTriangles method to the IRenderer and DirectXEngine class
Added line rendering support to the IMeshRenderer & DirectXMeshRenderer class
Fixed DirectXMeshRenderer default transform not being the default transform matrix
Fixed CustomMesh not being destroyed if it is replaced by an other mesh

LeeVangraefschepe commented on Graphics/DirectX/DirectXEngine.cpp:
```diff
@@ -342,6 +360,12 @@ void leap::graphics::DirectXEngine::Draw()
+		if (!m_DebugDrawings.GetIndexBuffer().empty())
+		{
+			m_pDebugRenderer->LoadMesh(m_DebugDrawings);
+			m_DebugDrawings = {}; // Reset debug renderer
```
Maybe create a reset/clear function for CustomMesh. Now every frame the vector will get recreated and resized along the frame calls. Problem with not recreating is that the vector stays large if many calls happend.

LeeVangraefschepe commented on LeapEngine/Leap.cpp:
```diff
@@ -106,14 +122,18 @@ void leap::LeapEngine::Run(int desiredFPS)
+        do
+        {
+            curFrameTimeNs = (std::chrono::high_resolution_clock::now() - currentTime).count();
+        } while (frameTimeNs - curFrameTimeNs > 0);
```
This will set the CPU thread on 100% for limited frames. Fixes it for now but should be looked into.

LeeVangraefschepe commented on Physics/Data/Rigidbody.h:
```diff
@@ -0,0 +1,109 @@
+		const glm::vec3& GetTranslation() const { return m_Translation; }
+		const glm::quat& GetRotation() const { return m_Rotation; }
+		const glm::quat& GetRotationDelta() const { return m_RotationDelta; }
+		const std::vector<Constraint> GetConstraints() { return m_Constraints; }
```
Returning const vector? Should prob be const &

AtlantiaKing committed:
Fixed typo (DrawTriangles should be called DrawLines)

LeeVangraefschepe commented on Physics/PhysX/PhysXMaterial.cpp:
```diff
@@ -0,0 +1,47 @@
+#include <PxPhysics.h>
+#include <PxMaterial.h>
+
+#include <Debug.h>
```
Debug include is unused

LeeVangraefschepe commented on Physics/PhysX/PhysXEngine.h:
```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include "../Interfaces/IPhysics.h"
+#include "../Interfaces/IPhysicsObject.h"
```
Unused include

LeeVangraefschepe commented on Physics/PhysX/PhysXObject.cpp:
```diff
@@ -0,0 +1,267 @@
+#include <PxRigidActor.h>
+#include <PxRigidStatic.h>
+#include <PxRigidDynamic.h>
+#include <PxRigidActor.h>
```
Included twice

LeeVangraefschepe commented on Physics/PhysX/PhysXObject.cpp:
```diff
@@ -0,0 +1,267 @@
+#include "PhysXScene.h"
+
+#include "../Data/Rigidbody.h"
+#include "../Data/ForceMode.h"
```
Unused include

LeeVangraefschepe commented on Physics/PhysX/PhysXObject.cpp:
```diff
@@ -0,0 +1,267 @@
+#include <PxRigidStatic.h>
+#include <PxRigidDynamic.h>
+#include <PxRigidActor.h>
+#include <PxScene.h>
```
Unused include

LeeVangraefschepe commented on Physics/PhysX/PhysXObject.cpp:
```diff
@@ -0,0 +1,267 @@
+	if (!IsValid()) static_cast<PhysXScene*>(pScene)->RemoveActor(m_pActor);
+}
+
+void leap::physics::PhysXObject::Apply(const std::function<void(void*, const glm::vec3&, const glm::quat&)>& setFunc, const std::function<std::pair<const glm::vec3&, const glm::quat&>(void*)> getFunc)
```
Function can be made const

LeeVangraefschepe commented on Physics/PhysX/PhysXObject.cpp:
```diff
@@ -0,0 +1,267 @@
+{
+	if (m_pRigidbody == nullptr)
+	{
+		auto transformPair{ getFunc(m_pOwner) };
```
Can be made const

LeeVangraefschepe commented on Physics/PhysX/PhysXObject.h:
```diff
@@ -0,0 +1,65 @@
+
+#include "../Interfaces/IPhysicsObject.h"
+
+#include <Subject.h>
```
Unused include

LeeVangraefschepe commented on Physics/PhysX/PhysXScene.cpp:
```diff
@@ -0,0 +1,55 @@
+	for (physx::PxU32 i = 0; i < rb.getNbLines(); i++)
+	{
+		const physx::PxDebugLine& line = pLines[i];
+		debugDrawings.emplace_back(std::make_pair(
```
std::make_pair is not needed for emplace_back

LeeVangraefschepe commented on Physics/PhysX/PhysXScene.cpp:
```diff
@@ -0,0 +1,55 @@
+	return debugDrawings;
+}
+
+void leap::physics::PhysXScene::AddActor(physx::PxRigidActor* pActor)
```
Function can be const

LeeVangraefschepe commented on Physics/PhysX/PhysXScene.cpp:
```diff
@@ -0,0 +1,55 @@
+	m_pScene->addActor(*pActor);
+}
+
+void leap::physics::PhysXScene::RemoveActor(physx::PxRigidActor* pActor)
```
Function can be const

LeeVangraefschepe commented on Physics/PhysX/PhysXShapes.h:
```diff
@@ -0,0 +1,80 @@
+		physx::PxShape* m_pShape{};
+	};
+
+	class PhysXBoxShape final : public IPhysXShape
```
Missing rule of 5 (destructor is not default)

LeeVangraefschepe commented on Physics/PhysX/PhysXShapes.h:
```diff
@@ -0,0 +1,80 @@
+		virtual glm::vec3 GetRelativePosition() override;
+	};
+
+	class PhysXSphereShape final : public IPhysXShape
```
Missing rule of 5 (destructor is not default)

LeeVangraefschepe commented on Physics/PhysX/PhysXShapes.h:
```diff
@@ -0,0 +1,80 @@
+		virtual glm::vec3 GetRelativePosition() override;
+	};
+
+	class PhysXCapsuleShape final : public IPhysXShape
```
Missing rule of 5 (destructor is not default)

LeeVangraefschepe commented on Physics/PhysX/PhysXSimulationCallbacks.cpp:
```diff
@@ -0,0 +1,59 @@
+#include "PhysXSimulationCallbacks.h"
+
+#include <Debug.h>
+#include <thread>
```
Unused include

LeeVangraefschepe commented on Physics/PhysX/PhysXSimulationFilterCallback.cpp:
```diff
@@ -0,0 +1,41 @@
+#include "PhysXSimulationFilterCallback.h"
+
+#include <Debug.h>
```
Unused include

LeeVangraefschepe commented on UnnamedAdventureGame/CMakeLists.txt:
```diff
@@ -9,7 +9,9 @@ add_executable(UnnamedAdventureGame
-)
+    "Components/PrintCollision.cpp"
+	"Components/ApplyForces.cpp" 
+	"Components/ColliderScaler.cpp" "Components/PrintVelocity.h" "Components/PrintVelocity.cpp")
```
Header in CMake and on same line

LeeVangraefschepe commented on Utils/Quaternion.h:
```diff
@@ -0,0 +1,26 @@
+		// Quaternion is constructable because it is NOT the object representation for a quaternion, 
+		//		use glm::quat for this
+		// This class exists to give some helper functions to generate correct quaternions/eulerangles
+		Quaternion() = default;
```
Constructor can be deleted

LeeVangraefschepe reviewed (changes requested):
Applied feedback (some you already did live)
Request again after the components have the physics calls to derive from.

AtlantiaKing committed:
Moved OnCollision/OnTrigger callbacks to the component class

Added OnCollision/OnTrigger methods to the Component class
Added OnCollision/OnTrigger methods to the GameObject class which calls the component versions of the methods
PhysicsSync now calls the gameobject oncollision/ontrigger methods instead of the collider methods

AtlantiaKing committed:
Deleted constructor for Quaternion class

Instead of having private Quaternion constructor/destructor, I changed it so the class deletes its constructor

AtlantiaKing committed:
Added clear method to the CustomMesh class

Instead of having to recreate a CustomMesh object every frame, it can now be cleared which keeps the capacity

AtlantiaKing committed:
Applied code review feedback

Const correctness in PhysXObject & PhysXScene
Removed redundant includes in PhysXEngine, PhysXMaterial, PhysXObject and PhysXSimulationCallbacks
GetConstraints function in the Rigidbody class now returns a const reference instead of a copy
Removed make_pair call for more readable code in the GetDebugDrawings function in the PhysXScene
Refactored IPhysxShape so it deletes the PxShape in its own destructors instead of the derived destructor and made the GetShape method non-virtual.
Cleaned unag CMakeLists

AtlantiaKing committed:
Merge branch 'main' into Physics-integration

AtlantiaKing committed:
Cleaned up Leap class

SetEnabledDebugDrawing is now being called in the main.cpp instead of the Leap.cpp
Removed m_Renderer variable in the Leap class

LeeVangraefschepe reviewed (approved):
Feedback has been applied
